didman/didman.go

```diff
@@ -66,25 +73,49 @@ func (d *didman) Name() string {
 
 func (d *didman) AddEndpoint(id did.DID, serviceType string, u url.URL) error {
 	logging.Log().Debugf("Adding endpoint (did: %s, type: %s, url: %s)", id.String(), serviceType, u.String())
-	doc, meta, err := d.docResolver.Resolve(id, nil)
-	if err != nil {
-		return err
+	err := d.addService(id, serviceType, u.String(), nil)
+	if err == nil {
+		logging.Log().Infof("Endpoint added (did: %s, type: %s, url: %s)", id.String(), serviceType, u.String())
 	}
+	return err
+}
 
-	// check for duplicate service type
-	for _, s := range doc.Service {
-		if s.Type == serviceType {
-			return ErrDuplicateService
+func (d *didman) AddCompoundService(id did.DID, serviceType string, references map[string]ssi.URI) error {
+	logging.Log().Debugf("Adding compound service (did: %s, type: %s, references: %v)", id.String(), serviceType, references)
+
+	// Resolve and validate all service references: they must all resolve to a single endpoint URL
+	// Cache resolved DID documents because most of the time a compound service will refer the same DID document in all service references.
+	documents := make(map[string]*did.Document)
+	var err error
+	for _, serviceRef := range references {
```

Comment:
Resolving the referenced documents can have its own function, reduces complexity.